foresight/utils/handler_utils.py

```diff
@@ -0,0 +1,187 @@
+
+        Args:
+            executor (ForesightExecutor): Keeps functions for start and finish thundra plugins
+            api_key (str, optional): Thundra api key to send data rest collector. Defaults to None.
+        """
+        try:
+            import thundra 
+            already_configured = True if ConfigProvider.configs else False
+            thundra._set_thundra_for_test_env(already_configured)
+            EnvironmentSupport.init()
+            TestWrapperUtils(api_key=api_key, plugin_executor = executor)
+            TestRunnerSupport.start_test_run()
+        except Exception as e:
+            logger.error("Thundra couldn't initialized for test: {}".format(e))
+            pass
+
+
+    @classmethod
+    def test_teardown(cls):
+        """For now, TestRunnerSupport not support concurrency. Only one test suite info is kept there.
+        It should be changed for concurrent python test framework.
+        """
+        if (TestRunnerSupport.test_suite_execution_context and 
+            not TestRunnerSupport.test_suite_execution_context.completed):
+            HandlerUtils.finish_test_suite_span()
+        TestRunnerSupport.finish_current_test_run()
+
+
+    @classmethod
+    def start_test_suite_span(cls, test_suite_id, app_info):
+        """For now, TestRunnerSupport not support concurrency. Only one test suite info is kept there.
+        It should be changed for concurrent python test framework.
+        """
+        if not TestRunnerSupport.test_suite_execution_context:
+            test_wrapper_utils = TestWrapperUtils.get_instance()
+            context = test_wrapper_utils.create_test_suite_execution_context(test_suite_id)
+            ExecutionContextManager.set(context)
+            test_wrapper_utils.change_app_info(app_info)
+            TestRunnerSupport.set_test_suite_application_info(app_info)
+            TestRunnerSupport.set_test_suite_execution_context(context)
+            test_wrapper_utils.before_test_process(context)
+
+
+    @classmethod
+    def start_before_all_span(cls, app_info, span_tags= None):
+        return cls.create_span(cls.TEST_BEFORE_ALL_OPERATION_NAME, app_info, span_tags)
+
+
+    @classmethod
+    def finish_before_all_span(cls, scope):
+        cls.finish_span(scope, TestRunnerTags.TEST_BEFORE_ALL_DURATION)
+
+
+    @classmethod
+    def start_after_all_span(cls, app_info, span_tags):
+        """after all executed after test cases. That is why context should be getting from TestRunnerSupport and
+        set into ExecutionContextManager.
+        """
+        context = TestRunnerSupport.test_suite_execution_context
+        ExecutionContextManager.set(context)
+        return cls.create_span(cls.TEST_AFTER_ALL_OPERATION_NAME, app_info, span_tags)
+
+
+    @classmethod
+    def finish_after_all_span(cls, scope):
+        cls.finish_span(scope, TestRunnerTags.TEST_AFTER_ALL_DURATION)
+
+
+    @staticmethod
+    def finish_test_suite_span():
```

Comment:
where we close test suite span?

Reply:
It is also closed into plugin.py when next test case is None or current test case and next test case's module are not equal or session is finished.